examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,208 @@
+## Example for calculating and plotting the Probability of Improvement
+
+# This example demonstrates how to use the
+# {func}`register_hooks <baybe.utils.basic.register_hooks>` utility
+# to calculate and analyse the probability of improvement (PI) evolving during iterations:
+# * We define a hook that is compatible with the general
+#   {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`
+#   interface,
+# * attach it to a recommender,
+# * and watch it take action.
+
+
+### Imports
+
+import os
+import sys
+import warnings
+from pathlib import Path
+from types import MethodType
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+from botorch.test_functions.synthetic import Hartmann
+from matplotlib.collections import PolyCollection
+from matplotlib.figure import Figure
+from scipy.stats import gaussian_kde
+
+from baybe.acquisition import ProbabilityOfImprovement
+from baybe.campaign import Campaign
+from baybe.objectives.base import Objective
+from baybe.objectives.single import SingleTargetObjective
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import (
+    BotorchRecommender,
+    RandomRecommender,
+    TwoPhaseMetaRecommender,
+)
+from baybe.searchspace import SearchSpace, SearchSpaceType
+from baybe.surrogates import GaussianProcessSurrogate
+from baybe.targets import NumericalTarget
+from baybe.utils.basic import register_hooks
+from baybe.utils.botorch_wrapper import botorch_function_wrapper
+from baybe.utils.dataframe import to_tensor
+
+### Parameters for a full simulation loop
+
+# For the full simulation, we need to define some additional parameters. These are the number of experiments to be conducted per run, the batch size, the dimension and the points per dimension.
+
+SMOKE_TEST = "SMOKE_TEST" in os.environ
+N_DOE_ITERATIONS = 2 if SMOKE_TEST else 7
+BATCH_SIZE = 1
```

Comment:
@AdrianSosic @Scienfitz @RimRihana : As mentioned in the chat, I have added some random noise to avoid the issues of numerical stability, making it possible to use `batch_size=1`. However, for some reason that I do not undestand, the first recommendation made by the `BotorchRecommender` has POI of 0.5 everywhere in the case of `batch_size=1`. I do not know what the reason for this is, but we need to discuss how to handle this since this makes the plots basically unusable as there is a huge peak for the first iteration and one does not see any additional curves. My different suggestions, please comment on them:
1. Use batch_size>=2.
2. Do not plot the very first POI
3. Investigate the issue in detail

While 3. is something that we should definitely do, I am not sure whether we should do it now, since this example should be merged this week. Both 1. and 2. are fine for me, while 1. is my favorite. How about you?

Reply:
Quick answer: poi of 0.5 makes perfectly sense for `batch_size=1` since it means you only have one point collected in the first iteration. As long as you are using a stationary kernel, your posterior GP will be a constant with a fixed uncertainty around it, so all other points are considered equally promising. Or in layman terms: if you have only seen one data point, there is a 50:50 chance that your next data point will be better/worse

Reply:
Does that already solve your problem?

Reply:
sorry: the "fixed uncertainty around it" is wrong, because you'll have a reduced uncertainty in the proximity of your point, but the argument still holds since the GP is constant and thus 50% of the probability mass will lie above the point and 50% below, since symmetric Gaussian distribution

Reply:
For me that argument makes perfect sense, thanks for the information! That was the input I needed to understand why this happens :)
Still - the plots look very bad when using batch size of 1 since the first iteration just has a single peak at POI=0.5, making all other iterations barely visible - so how should we deal with that? With that explanation, my favorite shifts towards option 2 (together with a comment why the first one is nor included in the plot)

Reply:
I've no strong opinion here, happy with whatever works. Probably I'd simply pick the option which is easiest to explain in the text 🙃 Here yet another option, but I don't know if better or not: you can also set the switching point of the meta recommender to a higher number (e.g. 2) and then your botorch recommender will not be executed in that setting in the first place

Reply:
honestly, the best solution then is `batchsize=2` and reiterate to find a slightly better function/problem so the curve evolution is prettier / slower

Reply:
Percect. THen this can be resolved once a suitable example for `batch_size=2` was found.

Reply:
For me that is fine. The shape for iteration 0 is weird, but acceptable imo. What about @Scienfitz and @AdrianSosic ?

Reply:
i think this has been solved by using a linspace for x axis points
also a good random seed has been found to create a nice looking plot
resolving